src/compiler/checker.ts

```diff
@@ -4583,7 +4583,10 @@ namespace ts {
                 let requireOptionalProperties = relation === subtypeRelation && !(source.flags & TypeFlags.ObjectLiteral);
                 for (let targetProp of properties) {
                     let sourceProp = getPropertyOfType(source, targetProp.name);
+
                     if (sourceProp !== targetProp) {
+                        // sourceProp !== targetProp -- ie: source and target have distinct declarations with the same name

```

Comment:
Not necessarily. They may be declared in the same place, but let's say they are in a generic type. They will be two different instantiations of the same declared property:
```ts
interface Foo<T> {
    prop: T;
}
var a: Foo<string>;
var b: Foo<number>;
a = b; // will compare two different props declared at the same place.
```